### PR TITLE
treewide: Update to golang 1.21

### PR DESCRIPTION
**Pull request by mqasimsarfraz**
As discussed in the comment [here](https://github.com/inspektor-gadget/inspektor-gadget/pull/2195#discussion_r1381991808) the pros/cons of updating to golang 1.21:

>  I think 1.21 with its maps and slices packages and the new min() and max() funcs alone are worth the upgrade as we use those things constantly. And this is timed just after a release, so there's plenty of time ahead of us to make sure it's running well. So my opinion is: go for it.
>
>  The release notes state that Go 1.21 no longer runs on Win 7 & 8 and older MacOS versions, but I don't know if the binaries it produces are affected as well - we should check for that and maybe add that to the release notes.

We can continue the reviews and feedback separately from https://github.com/inspektor-gadget/inspektor-gadget/pull/2195

### CODE REVIEW

File: cmd/common/utils/flags.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"golang.org/x/exp/slices"
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply cmd/common/utils/flags.go:26]: Good catch, [slices](https://github.com/golang/go/commit/1e9925f461c54f5fdc100f9f23916b3df81262ce) has already been added to standard packages. I will update the PR!

But yes [`exp`](https://pkg.go.dev/golang.org/x/exp#section-readme) is coming from experimental.

File: cmd/common/utils/flags.go [hunk 2]
```diff
@@ -18,6 +18,7 @@ import (
 	"errors"
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply cmd/common/utils/flags.go:18]: Agree :D I update it now!

**Comment by mqasimsarfraz** [comment]
I will merge this by EOD tomorrow in case no one has any concerns!

**Review by flyth — APPROVED** [review]
Thanks for addressing this, looking forward to using the new packages / funcs :)

Just a note regarding the `go.mod` - unsure if it has got something to do with the Go version bump or maybe was just leaked from another branch?

File: go.sum [hunk 3]
```diff
@@ -57,6 +58,7 @@ github.com/Microsoft/go-winio v0.6.1/go.mod h1:LRdKpFKfdobln8UmuiYcKPot9D2v6svN5
 github.com/Microsoft/hcsshim v0.12.0-rc.0 h1:wX/F5huJxH9APBkhKSEAqaiZsuBvbbDnyBROZAqsSaY=
 github.com/Microsoft/hcsshim v0.12.0-rc.0/go.mod h1:rvOnw3YlfoNnEp45wReUngvsXbwRW+AFQ10GVjG1kMU=
 github.com/armon/go-socks5 v0.0.0-20160902184237-e75332964ef5 h1:0CwZNZbxp69SHPdPJAN/hZIm0C4OItdklCFmMRWYpio=
+github.com/armon/go-socks5 v0.0.0-20160902184237-e75332964ef5/go.mod h1:wHh0iHkYZB8zMSxRWpUBQtwG5a7fFgvEO+odwuTv2gs=
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply go.sum:61]: It seems all the new entries have `/go.mod` in the version so they aren't new modules but perhaps golang 1.21 is ensuring that we have checksum for go.mod for each module. Looks related to this [change](https://go-review.googlesource.com/c/go/+/489075) but I am not sure ?

[the remark this answers is not in the document]
Comment (flyth) [reply go.sum:61]: Thanks for the clarification, it looks like that is the reason.

**Review by blanquicet — APPROVED** [review]

**Comment by mqasimsarfraz** [comment]
Thanks for the reviews!

**Comment by mqasimsarfraz** [comment]
After reading some of the discussions [[1]](https://github.com/atc0005/go-ci/issues/1188) [[2]](https://github.com/opencontainers/image-spec/pull/1114#pullrequestreview-1639441396) regarding `GOTOOLCHAIN` in golang 1.21. I wanted to check if we should set `GOTOOLCHAIN=local` in our CI/Makefile. It has been handled for golang [dockerfiles](https://github.com/docker-library/golang/pull/491) so only effects when we build directly.

The concern is that, in future the golang version in the [CI](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/.github/workflows/inspektor-gadget.yml#L5) might not be the one used to build binaries because [gotoolchain](https://go.dev/doc/toolchain) will automatically download and build with newer version if the modules requires it. Do we foresee any problem with this?

**Pull request merged**